src/main/java/frc/trigon/robot/subsystems/swerve/Swerve.java

```diff
@@ -86,31 +112,53 @@ public boolean atPose(MirrorablePose2d pose2d) {
+        return atTranslationPosition(RobotContainer.POSE_ESTIMATOR.getCurrentEstimatedPose().getX(), xAxisPosition, currentXAxisVelocity);
     }
 
     public boolean atYAxisPosition(double yAxisPosition) {
         final double currentYAxisVelocity = getFieldRelativeVelocity().vyMetersPerSecond;
-        return atTranslationPosition(RobotContainer.POSE_ESTIMATOR.getCurrentPose().getY(), yAxisPosition, currentYAxisVelocity);
+        return atTranslationPosition(RobotContainer.POSE_ESTIMATOR.getCurrentEstimatedPose().getY(), yAxisPosition, currentYAxisVelocity);
     }
 
     public boolean atAngle(MirrorableRotation2d angle) {
-        final boolean atTargetAngle = Math.abs(angle.get().minus(RobotContainer.POSE_ESTIMATOR.getCurrentPose().getRotation()).getDegrees()) < SwerveConstants.ROTATION_TOLERANCE_DEGREES;
+        final boolean atTargetAngle = Math.abs(angle.get().minus(RobotContainer.POSE_ESTIMATOR.getCurrentEstimatedPose().getRotation()).getDegrees()) < SwerveConstants.ROTATION_TOLERANCE_DEGREES;
         final boolean isAngleStill = Math.abs(getSelfRelativeVelocity().omegaRadiansPerSecond) < SwerveConstants.ROTATION_VELOCITY_TOLERANCE;
         Logger.recordOutput("Swerve/AtTargetAngle/isStill", isAngleStill);
         Logger.recordOutput("Swerve/AtTargetAngle/atTargetAngle", atTargetAngle);
         return atTargetAngle/* && isAngleStill*/;
     }
 
-    public SwerveModulePosition[] getWheelPositions() {
-        final SwerveModulePosition[] swerveModulePositions = new SwerveModulePosition[swerveModules.length];
+    public double[] getDriveWheelPositionsRadians() {
+        final double[] swerveModulesPositions = new double[swerveModules.length];
         for (int i = 0; i < swerveModules.length; i++)
-            swerveModulePositions[i] = swerveModules[i].getOdometryPosition(swerveModules[i].getLastOdometryUpdateIndex());
-        return swerveModulePositions;
+            swerveModulesPositions[i] = swerveModules[i].getDriveWheelPosition();
+        return swerveModulesPositions;
+    }
+
+    public Rotation2d getDriveRelativeAngle() {
+        final Rotation2d currentAngle = RobotContainer.POSE_ESTIMATOR.getCurrentEstimatedPose().getRotation();
+        return Mirrorable.isRedAlliance() ? currentAngle.rotateBy(Rotation2d.fromDegrees(180)) : currentAngle;
+    }
+
+    public void runWheelRadiusCharacterization(double omegaRadiansPerSecond) {
+        selfRelativeDrive(new ChassisSpeeds(0, 0, omegaRadiansPerSecond));
     }
 
-    public void runWheelRadiusCharacterization(double omegaRadsPerSec) {
-        selfRelativeDrive(new ChassisSpeeds(0, 0, omegaRadsPerSec));
+    /**
+     * Drives the swerve with the given chassis speeds and feedforwards, relative to the robot's frame of reference.
+     *
+     * @param chassisSpeeds the target chassis speeds
+     * @param feedforwards  the target feedforwards
+     */
+    public void selfRelativeDrive(ChassisSpeeds chassisSpeeds, DriveFeedforwards feedforwards) {
```

Comment:
no reason for the feedforwards to be here
we can create a new function specifically for PP when you figure out how to use PP, but there's no reason to add a mostly redundant variable